### PR TITLE
Antimagic works again

### LINKED ISSUE #42322: antimagic doesnt protect from arcane bolts
had holymelons in pockets and it didnt protect me from arcane bolts
100087

### PR DESCRIPTION
**Pull request by 4dplanner**
## Changelog
:cl:
fix: antimagic works again
/:cl:
Closes #42322

### CODE REVIEW

**Comment by necromanceranne** [comment]
Speed merge plz

File: code/datums/components/anti_magic.dm [hunk 1]
```diff
@@ -33,12 +33,14 @@
 	UnregisterSignal(user, COMSIG_MOB_RECEIVE_MAGIC)
 
 /datum/component/anti_magic/proc/protect(datum/source, mob/user, _magic, _holy, major, self, list/protection_sources)
-	if(((_magic && magic) || (_holy && holy)) && (self && blocks_self))
+	if(((_magic && magic) || (_holy && holy)) && (!self || !blocks_self))
```
Comment (Tlaltecuhtli) [inline code/datums/components/anti_magic.dm]: :smoking:

Comment (4dplanner) [reply code/datums/components/anti_magic.dm]: Fuck there's an extra ! in there, will fix in a bit

Comment (4dplanner) [reply code/datums/components/anti_magic.dm]: Fuck there's an extra ! in there, will fix in a bit

Comment (4dplanner) [reply code/datums/components/anti_magic.dm]: Fuck there's an extra ! in there, will fix in a bit

Comment (Therapist1) [reply code/datums/components/anti_magic.dm]: an extra you say?

Comment (4dplanner) [reply code/datums/components/anti_magic.dm]: You saw nothing

**Comment by MrDoomBringer** [comment]
Wtf i like 4dplanner now?

**Comment by Shardj** [comment]
I'm confused, how does this nerf botany?

**Comment by 4dplanner** [comment]
> I'm confused, how does this nerf botany?

It buffs it :'(

File: code/datums/components/anti_magic.dm [hunk 2]
```diff
@@ -33,12 +33,14 @@
 	UnregisterSignal(user, COMSIG_MOB_RECEIVE_MAGIC)
 
 /datum/component/anti_magic/proc/protect(datum/source, mob/user, _magic, _holy, major, self, list/protection_sources)
-	if(((_magic && magic) || (_holy && holy)) && (self && blocks_self))
+	if(((_magic && magic) || (_holy && holy)) && (!self || blocks_self))
 		protection_sources += parent
-		reaction.Invoke(user, major)
+		if(reaction)
```
Comment (Cyberboss) [inline code/datums/components/anti_magic.dm]:
```suggestion
```

File: code/datums/components/anti_magic.dm [hunk 3]
```diff
@@ -33,12 +33,14 @@
...
+			reaction.Invoke(user, major)
```
Comment (Cyberboss) [inline code/datums/components/anti_magic.dm]:
```suggestion
		reaction?.Invoke(user, major)
```

File: code/datums/components/anti_magic.dm [hunk 4]
```diff
@@ -33,12 +33,14 @@
...
 		if(major)
 			charges--
 			if(charges <= 0)
-				expire.Invoke(user)
+				if(expire)
```
Comment (Cyberboss) [inline code/datums/components/anti_magic.dm]:
```suggestion
```

File: code/datums/components/anti_magic.dm [hunk 5]
```diff
@@ -33,12 +33,14 @@
...
+					expire.Invoke(user)
```
Comment (Cyberboss) [inline code/datums/components/anti_magic.dm]:
```suggestion
				expire?.Invoke(user)
```

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by fizzle7** [comment]
But why was it null? This obscures the problem. Also, is ? really necessary? Just makes it harder to read.

**Comment by 4dplanner** [comment]
@fizzle7 There are two changes here - one checks for callbacks being null, which is a perfectly fine way of signalling there is no callback, and the second is a logic error. Neither is being "obscured".

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**